### PR TITLE
JSONify range_with_even_chance_of_good_hit

### PR DESCRIPTION
**Pull request by SlackingSleeper**
#### Summary

SUMMARY: Infrastructure "Adds JSON capability to range_with_even_chance_of_good_hit."
#### Purpose of change

Previously range_with_even_change_of_good_hit had a hardcoded array to pass over. With the change you can update its values simply by running synthetic_range_test instead of having to change the hardcoded array.

#### Describe the solution

range_test in ranged_balance writes its data to ./data/json/hit_range.json which is loaded when loading a world ingame. When range_with_even_chance_of_good_hit is called it checks this data or defaults to the old data.

### CODE REVIEW

**Comment by Qrox** [comment]
Is this generated JSON file supposed to be used in the main game code? If not perhaps it's better to load it in the test only.

**Comment by SlackingSleeper** [comment]
It is used in item.cpp and npcmove.cpp.

File: tests/ranged_balance.cpp [hunk 1]
```diff
@@ -334,8 +337,19 @@ static void range_test( const std::array<double, 5> &test_thresholds )
             WARN( "No matching dispersion found" );
         } else {
             WARN( "Range: " << r << " Dispersion: " << found_dispersion );
+            data.push_back( found_dispersion );
         }
     }
+    if( write_data )
+    {
+        std::ofstream out_file;
```
Comment (BevapDin) [inline tests/ranged_balance.cpp]: Please consider using `write_to_file` instead of using `ofstream` directly.

Comment (SlackingSleeper) [reply tests/ranged_balance.cpp]: I'm getting a really weird error after implementing write_to_file. Did I do something wrong?

`.Terminated: data/mods//dda/../../json/hit_range.json: line 1:1: member not foun
d: type

{
^
"even_good":[1723,859,573,429,343,283,245,214,190,171,156,142,132,122,114,107,10
1,88,88,85,81,78,74,71,68,66,63,61,59,57,55,53,52,50,49,47,46,45,44,42,41,41,39,
39,38,37,36,35,35,34,33,33,32,31,31,30,30,29,29,28]}}}}}}}}}}}}}}}}}}}}}}}}}}}}}`

File: tests/ranged_balance.cpp [hunk 2]
```diff
@@ -334,8 +337,18 @@ static void range_test( const std::array<double, 5> &test_thresholds )
             WARN( "No matching dispersion found" );
         } else {
             WARN( "Range: " << r << " Dispersion: " << found_dispersion );
+            data.push_back( found_dispersion );
         }
     }
+    if( write_data )
+    {
+        write_to_file( "./data/json/hit_range.json", [&]( std::ostream & fsa ){
```
Comment (BevapDin) [inline tests/ranged_balance.cpp]: This writes a ".json" file into the data/json directory. The JSON loader (init.cpp) loads *all* ".json" files from that directory. And it expects those files to have objects with a "type" member.

This code does not write a "type" member, so the loader does not know how to load it.

I suggest adding a "type" like this `j_out.member("type", "hit_range");`

**Comment by Firestorm01X2** [comment]
@SlackingSleeper 

Just to clarify. This array does not affect dispersion calculation. 

**Changing it does not make guns more or less accurate.**

 It is used just for determining chance to even good hit. And this array is separated from actual accuracy calculation.
This data precalculated with test. **So this array just for dispalying chance itself.**

File: tests/ranged_balance.cpp (hunk 2 above)
Comment (SlackingSleeper) [reply tests/ranged_balance.cpp]: Added in the latest commit.

**Comment by Firestorm01X2** [comment]
Corrected my initial post. This array used to determine **range** of even good hit. Other things the same. This is separated from actual ranged calculation.

**Comment by Qrox** [comment]
It should probably also check the test results against `hit_range.json` in the repository and alert if they are different, so any change in the results in the future gets properly commited.

**Comment by SlackingSleeper** [comment]
> 
> 
> It should probably also check the test results against `hit_range.json` in the repository and alert if they are different, so any change in the results in the future gets properly commited.

How is this done? Also due to the nature of the test the data can vary ~1-2 dispersion each time it's run.

**Comment by Qrox** [comment]
I think you can check if the new dispersion data is close enough to the old one. Something like (I haven't tested)
```c++
REQUIRE( std::equal( data.begin(), data.end(), Creature::dispersion_for_even_chance_of_good_hit.begin(), Creature::dispersion_for_even_chance_of_good_hit.end(), []( const int a, const int b ) -> bool {
    return a > 0 && b > 0 && std::abs( static_cast<float>( a - b ) / b ) < 0.05;
} ) );
```
It checks whether every new dispersion value is within 5% range of the old value.

File: tests/ranged_balance.cpp [hunk 3]
```diff
@@ -334,6 +338,27 @@ static void range_test( const std::array<double, 5> &test_thresholds )
...
+        }
+    }
+    if( write_data )
+    {
+        if( std::equal( data.begin(), data.end(),
+                        Creature::dispersion_for_even_chance_of_good_hit.begin(),
+                        Creature::dispersion_for_even_chance_of_good_hit.end(),
+                        []( const int a, const int b ) -> bool {
+            return a > 0 && b > 0 && std::abs( static_cast<float>( a - b ) / b ) < 0.05;
+        } ) )
```
Comment (Qrox) [inline tests/ranged_balance.cpp]: Under this logic it only writes to the JSON file if the results *hasn't* changed significantly. What I meant was using the unit test's `REQUIRE()` macro to check the dispersion values, and fail the test if they changed considerably, so if the values are changed in the future, we can be automatically notified of the change.

Comment (SlackingSleeper) [reply tests/ranged_balance.cpp]: Should it use REQUIRE_FALSE?

Comment (Qrox) [reply tests/ranged_balance.cpp]: If you require `std::equal(...)` to be false, you'll get a test failure when the values have *not* changed. I don't think that's what we want here.

Comment (SlackingSleeper) [reply tests/ranged_balance.cpp]: That makes sense. I've noticed differences of more than 10% after running the test several times in a row. Would it be better to raise the limit to something like 15% or to accept some false-positives?

Comment (SlackingSleeper) [reply tests/ranged_balance.cpp]: That makes sense. I've noticed differences of more than 10% after running the test several times in a row. Would it be better to raise the limit to something like 15% or to accept some false-positives?

Comment (Qrox) [reply tests/ranged_balance.cpp]: Bump it to 15% then. Sounds like 5% would cause too many false positives.

**Comment by kevingranade** [comment]
You need to run astyle over your changed code, bracket indentation doesn't match project style.

**Pull request merged**